doc/best-practices.md

```diff
@@ -118,43 +108,95 @@ class Foo {
 }
 ```
 
-### if you have a setter/getter that just directly sets/gets a value, consider not having it instead
+### Prefer calling functions to emitting events
 
-Don't
+Events obfuscate code flow and coupling and should only be used when the emitter of the event is truly decoupled from things wishing ot subscribe to the event. Prefer direct function calls when the emitter and subscriber are inherently coupled.
+
+Don't:
 ```js
-class Foo {
-  constructor() {
-    this._val = 1;
-  }
-    
-  get val() {
-      return this._val;
-  }
+function onKeyDown(e)
+  clearTimeout(keyTimout);
+  keyTimout = setTimeout(() => scene.emit("action_typing_ended"), 500);
+  scene.emit("action_typing_started");
+} 
 
-  set val(newVal) {
-      this._val = newVal;
+scene.addEventListener("action_typing_started", () => window.APP.hubChannel.beginTyping());
+scene.addEventListener("action_typing_ended", () => window.APP.hubChannel.endTyping());
+``` 
+
+Do:
+```js
+function onKeyDown(e)
+  clearTimeout(keyTimout);
+  keyTimout = setTimeout(() => window.APP.hubChannel.endTyping(), 500);
+  window.APP.hubChannel.beginTyping()
+} 
+``` 
+
+### Don’t fail silently.
+
+Always at least log a warning or error to the console when something unexpected happens
+
+### Avoid defensive coding
+
+Related to no failing silently; Do not guard against conditions that should not occur or indicate problems that are the responsibility of some other code. Adding unnecessary guards will force future readers of the code to have to try and figure out what cases that guard is intending to handle. Also, as with failing silently, unnecessary guards can lead to masking of serious issues.
+
+
+Don't:
+```js
+if(window.APP && window.APP.shouldAlwaysExist && window.APP.shouldAlwaysExist.doSomething) {
+  window.APP.shouldAlwaysExist.doSomething();
+}
```

Comment:
This reminded me of our increased use of optional chaining (`foo?.bar()`). Perhaps if we should watch out for that as a code smell too.

Reply:
I agree, I actually don't love it because it makes this sort of defensive coding easier to do and easier to sneak in. I definitely have used it though, and I will admit it makes code that legitimately needs to be doing that sort of checking much less verbose